### PR TITLE
Right fibrations

### PR DESCRIPTION
**Pull request by TOTBWF**
# Description

This PR defines right fibrations, characterizes their fibre categories, and shows that discrete fibrations are right fibrations.
It also proves some additional helper lemmas along the way.

### CODE REVIEW

File: src/Cat/Displayed/Cartesian/Discrete.lagda.md [hunk 1]
```diff
@@ -167,6 +171,38 @@ are unique, we have $f = g$.
...
+    where open Discrete-fibration disc
+```
+
+We can use this fact in conjunction with the fact that all fibres are thin to show
+that every vertical morphism in a discrete fibration is invertible.
+
+```agda
+  discrete→vertical-invertible
+    : Discrete-fibration
+    → (∀ {x} {x′ x″ : E.Ob[ x ]} → (f′ : E.Hom[ B.id ] x′ x″) → is-invertible↓ f′)
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Discrete.lagda.md:195]: What's with the extra parens around this type?

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+A [cartesian fibration] $\ca{E}$ is said to be a **right fibration** if every
+morphism in $\ca{E}$ is cartesian.
+
+```agda
+record Right-fibration : Type (o ⊔ ℓ ⊔ o′ ⊔ ℓ′) where
+  no-eta-equality
+  field
+    is-fibration
+      : Cartesian-fibration
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:39]:
```suggestion
    is-fibration : Cartesian-fibration
```

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    cartesian
+      : ∀ {x y} {f : Hom x y}
+      → ∀ {x′ y′} (f′ : Hom[ f ] x′ y′)
+      → Cartesian f f′
+
+  open Cartesian-fibration is-fibration public
+```
+
+One notable fact is that all of the vertical morphisms of a right
+fibrations are invertible. In other words, if $\ca{E}$ is a right
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:49]:
```suggestion
One notable fact is that every vertical morphism in a right
fibration is invertible. In other words, if $\ca{E}$ is a right
```

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+fibration, then all of the fibre categories are groupoids. This
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:50]:
```suggestion
fibration, then all of its fibre categories are groupoids. This
```

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+vertical-invertible+fibration→right-fibration fib vert-inv
+  .Right-fibration.is-fibration = fib
+vertical-invertible+fibration→right-fibration fib vert-inv
+  .Right-fibration.cartesian {x = x} {f = f} {x′ = x′} {y′ = y′} f′ = f-cart where
+    open Cartesian-fibration fib
+    open Cartesian-lift renaming (x′ to x-lift)
+```
+
+To see this, recall that cartesian morphisms are stable under
+vertical retracts. The cartesian lift $f^{*}$ of $f$ is obviously
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:80]: Should there be a link to "stable under vertical retracts" here?

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+cartesian, so it suffices to show that there is a vertical retract
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:81]: Should that be "a vertical retraction"?

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+$x^{*} \to x'$. To construct this retract, we shall factorize $f'$
+over $f \cdot id$; which yields a vertical morphism $i^{*} : x' \to x^{*}$.
+By our hypotheses, $i^{*}$ is invertible, and thus is a retract.
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:84]: That should definitely be "is a retraction", right? Or, a retract of what, I guess?

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    f-cart : Cartesian f f′
+    f-cart = cartesian-vert-retract-stable
+      (has-lift f y′ .cartesian)
+      (inverses[]→from-has-section[] i*-inv.inverses′)
+      factors
+```
+
+As a corollary, we get that all discrete fibrations are right fibrations.
+Intuitively this is true, as sets are 0-groupoids.
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:120]:
```suggestion
Intuitively, this is true, as sets are 0-groupoids.
```

File: src/Cat/Displayed/Cartesian/Right.lagda.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+## Fibred Functors and Right Fibrations
+
+As every map in a right fibration is cartesian, every displayed functor
+into a right fibration is automatically fibred.
+
+```agda
+functor+right-fibration→fibred
+  : ∀ {o₂ ℓ₂ o₂′ ℓ₂′}
```
Comment (plt-amy) [inline src/Cat/Displayed/Cartesian/Right.lagda.md:139]: nice 👍

File: src/index.lagda.md [hunk 10]
```diff
@@ -642,6 +642,9 @@ open import Cat.Displayed.Cartesian.Street
 open import Cat.Displayed.Cartesian.Discrete
 -- Discrete fibrations are presheaves
 
+open import Cat.Displayed.Cartesian.Right
+-- Fibrations of groupoids
```
Comment (plt-amy) [inline src/index.lagda.md:646]: Shouldn't this be 
```suggestion
-- Fibrations in groupoids
```
?

Comment (plt-amy) [reply src/index.lagda.md:646]: Thanks so much for updating the index though

**Review by plt-amy — COMMENTED** [review]
This and the previous PR are awesome stuff; however it's in my nature to complain about formatting and wording

File: src/Cat/Displayed/Cartesian/Discrete.lagda.md (hunk 1 above)
Comment (TOTBWF) [reply src/Cat/Displayed/Cartesian/Discrete.lagda.md:195]: Copy paste oversight!

File: src/Cat/Displayed/Cartesian/Right.lagda.md (hunk 9 above)
Comment (TOTBWF) [reply src/Cat/Displayed/Cartesian/Right.lagda.md:139]: 😎

File: src/index.lagda.md (hunk 10 above)
Comment (TOTBWF) [reply src/index.lagda.md:646]: Oh nice catch; I've been trying to get better at remembering lol

**Pull request merged**